### PR TITLE
core/shared/platform/nuttx: mock several APIs for libc-wasi

### PR DESCRIPTION
**Pull request by yamt**

### CODE REVIEW

**Comment by no1wudi** [comment]
Is it better to add these dummy implementations to NuttX ?

**Comment by yamt** [comment]
> Is it better to add these dummy implementations to NuttX ?

i don't see the point to make an OS provide non-working impl.

**Comment by no1wudi** [comment]
But once these API implemented, we must remove these code.

**Comment by yamt** [comment]
> But once these API implemented, we must remove these code.

yes. what's wrong with it?

**Comment by no1wudi** [comment]
I just think it's better to put the os implementation related code on the os side, especially these API are defined by POSIX.

**Comment by yamt** [comment]
> I just think it's better to put the os implementation related code on the os side, especially these API are defined by POSIX.

ideally, yes.
however, implementing openat family is not trivial.
i think this makes sense as a stop-gap solution.

File: core/shared/platform/nuttx/platform_internal.h [hunk 1]
```diff
@@ -44,6 +50,60 @@ typedef pthread_t korp_thread;
...
+linkat(int fd1, const char *path1, int fd2, const char *path2, int flag);
+int
+renameat(int fromfd, const char *from, int tofd, const char *to);
+int
+symlinkat(const char *target, int fd, const char *path);
+int
+unlinkat(int fd, const char *path, int flag);
+int
+utimensat(int fd, const char *path, const struct timespec ts[2], int flag);
+#define AT_SYMLINK_NOFOLLOW 0
```
Comment (no1wudi) [inline core/shared/platform/nuttx/platform_internal.h:87]: Should we wrap these define like O_DIRECTORY ?

Comment (yamt) [reply core/shared/platform/nuttx/platform_internal.h:90]: i made these depend on !defined(AT_FDCWD)

**Comment by xiaoxiang781216** [comment]
@no1wudi , @Donny9 has a patch which is almost finished. Maybe, we can boost the progress.

**Comment by yamt** [comment]
i dropped preadv/pwritev mock after https://github.com/apache/incubator-nuttx/pull/6168

**Comment by yamt** [comment]
> @no1wudi , @Donny9 has a patch which is almost finished. Maybe, we can boost the progress.

i guess we can revert (part of?) this PR when/if his patch gets ready and merged to nuttx.

File: core/shared/platform/nuttx/platform_internal.h [hunk 2]
```diff
@@ -50,6 +56,56 @@ typedef pthread_t korp_thread;
...
+/*
+ * NuttX doesn't have O_DIRECTORY or directory open.
+ * REVISIT: maybe this is safer to be disabled at higher level.
+ */
+#if !defined(O_DIRECTORY)
+#define O_DIRECTORY 0
+#endif
+
+#if !defined(O_NOFOLLOW)
+#define O_NOFOLLOW 0
```
Comment (xiaoxiang781216) [inline core/shared/platform/nuttx/platform_internal.h:68]: ditto

**Comment by yamt** [comment]
@xiaoxiang781216 do you want to keep this blocked until @Donny9's work gets ready?

**Comment by xiaoxiang781216** [comment]
Let's merge it first? @Donny9 is busy on other task.

**Comment by yamt** [comment]
@no1wudi do you have any concern?

**Comment by no1wudi** [comment]
It's LGTM now, and once these APIs implemented on NuttX side, we must sync the changes to WAMR.

**Comment by wenyongh** [comment]
> It's LGTM now, and once these APIs implemented on NuttX side, we must sync the changes to WAMR.

@yamt @no1wudi Shall we merge the PR?

**Comment by yamt** [comment]
> > It's LGTM now, and once these APIs implemented on NuttX side, we must sync the changes to WAMR.
> 
> @yamt @no1wudi Shall we merge the PR?

i prefer to merge it because it would make my work simpler.

**Comment by no1wudi** [comment]
Yes we can merge it, it works with current tagged NuttX release at least.

**Pull request merged**